### PR TITLE
Fix block scalar leading/trailing newlines

### LINKED ISSUE #253: YAML 1.2 spec: Example 8.2. Block Indentation Indicator
When I try to parse example 8.2 at http://www.yaml.org/spec/1.2/spec.html#Block
I got:

```
[ 'detected\n',
  '\n\n# detected\n',
  '\n explicit\n',
  '\n\t\ndetected\n' ]
```

But according to the canonical form last two lines must not get a leading '\n':

```
%YAML 1.2

---
!!seq [
  !!str "detected\n",
  !!str "\n\n# detected\n",
  !!str "·explicit\n",
  !!str "\t·detected\n",
]
```

### LINKED ISSUE #276: construct-string-types test fails
See comment in https://github.com/nodeca/js-yaml/pull/274

This should be fixed before release.

### PR DESCRIPTION
**Pull request by aepsilon**
Fixes #253: loader should not add an extra newline at the beginning of block scalars that have an explicit indentation indicator.
Cause: "Detected indent" was assumed logically equivalent to "did read non-empty line".
This no longer holds with an indentation indicator: indent is detected before any content is read.

Fixes #276: loader should not remove a trailing newline for block scalars with "+" chomping.
Cause: "+" chomping didn't account for the last content newline before the trailing empty lines.

---

I'm not 100% familiar with the loader so there may be some missed edge cases.
So far it's passed every test I've tried.
@dervus and @puzrin, I hope this helps. Let me know if you find something.

### CODE REVIEW

**Comment by aepsilon** [comment]
Added a hotfix for double-quoted.
I was too focused on the other styles to notice, sorry. There's now a test case to cover it.

**Comment by dervus** [comment]
Sorry for so delayed reply!

Could you make `didReadContent` a Boolean variable? You use it for `+1` operation anyway, and using integers in Boolean expressions is against our coding style.

Also, this fixes only first part of #253, am I right?

**Comment by puzrin** [comment]
if `didReadContent` become boolean, equation will be `common.repeat('\n', didReadContent ? emptyLines + 1 : emptyLines)`

**Comment by aepsilon** [comment]
I prefer boolean too for style. Originally I wrote it like you said.
But it turns out auto-converting boolean to number is slow in V8, so I changed it for performance.
The added benchmarks bear this out.

There's another neat solution though: represent '\n' directly rather than converting it from 1 or true each time. The new commit uses this method to avoid almost all conversions and branching. It has the best performance and, I believe, good readability. What do you think?

**Benchmarks**

    1/0 as boolean and number (old commit)
    Sample: scalar_block_folded.yaml (2785 characters)
     > current x 39,255 ops/sec ±1.54% (76 runs sampled)
    Sample: scalar_block_literal.yaml (2785 characters)
     > current x 37,874 ops/sec ±0.78% (86 runs sampled)

    true/false as boolean and number
    Sample: scalar_block_folded.yaml (2785 characters)
     > current x 38,813 ops/sec ±1.23% (79 runs sampled)
    Sample: scalar_block_literal.yaml (2785 characters)
     > current x 33,871 ops/sec ±1.00% (85 runs sampled)

    ternary (didReadContent ? emptyLines + 1 : emptyLines))
    Sample: scalar_block_folded.yaml (2785 characters)
     > current x 39,068 ops/sec ±0.93% (78 runs sampled)
    Sample: scalar_block_literal.yaml (2785 characters)
     > current x 35,470 ops/sec ±0.89% (87 runs sampled)

    '\n'/'' (new commit)
    Sample: scalar_block_folded.yaml (2785 characters)
     > current x 38,504 ops/sec ±0.99% (75 runs sampled)
    Sample: scalar_block_literal.yaml (2785 characters)
     > current x 38,513 ops/sec ±0.61% (86 runs sampled)

---

For part 2 of #253, it appears the spec has an error.

`' \t'` is a more-indented line according to the grammar, since `\t` is `s-white`.
> [177] s-nb-spaced-text(n) ::= s-indent(n) s-white nb-char*

(Rule [175] for folded lines doesn't match)
Therefore join with a newline.

LibYAML 0.1.5 agrees.
It throws an error on the example, but if you specify indent with` >1`, it also produces `'\t\ndetected\n'`.

File: lib/js-yaml/loader.js [hunk 1]
```diff
@@ -852,7 +850,7 @@ function readBlockScalar(state, nodeIndent) {
 
       // Just one line break - perceive as the same line.
       } else if (emptyLines === 0) {
-        if (didReadContent) { // i.e. only if we have already read some scalar content.
+        if (contentNewline) { // i.e. only if we have already read some scalar content.
```
Comment (puzrin) [inline lib/js-yaml/loader.js]: `if (contentNewline.length)` ?

Comment (aepsilon) [inline lib/js-yaml/loader.js]: Or `!== ''`. I'd expect all of them to have about the same performance, since the type of `contentNewline` is fully determined as non-null string at compile time.

Comment (puzrin) [inline lib/js-yaml/loader.js]: I mean, we avoid indirect type casts and string operations when possible.

Comment (aepsilon) [inline lib/js-yaml/loader.js]: Sure

**Comment by dervus** [comment]
It doesn't seem to really affect the benchmarks in my opinion. The difference seem to be in the margin of error. I'd propose to stick Boolean solution @puzrin proposed.

**Comment by puzrin** [comment]
@dervus any more notes? Ok to squash and merge?

**Comment by dervus** [comment]
Yes, looks good to me. Thanks, @aepsilon.

**Comment by puzrin** [comment]
@aepsilon could you squash 2 last commits into the first one?

**Comment by aepsilon** [comment]
Done. :+1: Thanks for checking it over.

**Comment by puzrin** [comment]
Seems ok to release.

3.6.0 or 4.0.0? Can it break something?

**Comment by dervus** [comment]
Well, it's not an API change, though it changes dumper output. However I saw some NodeJS folks talking about changes of dump output. It was somewhere at PRs/issues external references.

I'd bet for minor version change.
